h2o-core/src/main/java/water/rapids/ast/prims/advmath/SpearmanCorrelation.java

```diff
@@ -1,72 +1,155 @@
+
+            correlationMatrix.vec(vecIdX)
+                    .set(vecIdY, spearman.getSpearmanCorrelationCoefficient());
+          }
+        } finally {
+          Scope.exit();
+        }
+      }
+    }
 
-public class AstSpearman extends AstPrimitive<AstSpearman> {
-  @Override
-  public int nargs() {
-    return 1 + 3; // Frame ID and ID of two numerical vectors to calculate SCC on.
+    return correlationMatrix;
   }
 
-  @Override
-  public String[] args() {
-    return new String[]{"frame", "first_column", "second_column"};
+  /**
+   * Compares two frames for their vecs being the same key-wise.
+   *
+   * @param frameX An instance of {@link Frame} to compare
+   * @param frameY A second instance of {@link Frame} to compare
+   * @return True if and only if the frames both contain the same vectors in the same order and quantity. Otherwise false.
+   */
+  private static boolean framesContainSameVecs(final Frame frameX, final Frame frameY) {
+    final Vec[] vecsX = frameX.vecs();
+    final Vec[] vecsY = frameY.vecs();
+
+    if (vecsX.length != vecsY.length) return false;
+
+    for (int i = 0; i < vecsX.length; i++) {
+      if (!vecsX[i]._key.equals(vecsY[i]._key)) return false;
+    }
+
+    return true;
   }
 
-  @Override
-  public Val apply(Env env, Env.StackHelp stk, AstRoot[] asts) {
-    final Frame originalUnsortedFrame = stk.track(asts[1].exec(env).getFrame());
-    final int vecIdX = originalUnsortedFrame.find(asts[2].exec(env).getStr());
-    final int vecIdY = originalUnsortedFrame.find(asts[3].exec(env).getStr());
-
-    try {
-      Scope.enter();
-      final SpearmanRankedVectors rankedVectors = rankedVectors(originalUnsortedFrame, vecIdX, vecIdY);
-      // Means must be calculated separately - those are not calculated for categorical columns in rollup stats.
-      final double[] means = calculateMeans(rankedVectors._x, rankedVectors._y);
-      final SpearmanCorrelationCoefficientTask spearman = new SpearmanCorrelationCoefficientTask(means[0], means[1])
-              .doAll(rankedVectors._x, rankedVectors._y);
-      return new ValNum(spearman.getSpearmanCorrelationCoefficient());
-    } finally {
-      Scope.exit();
+  /**
+   * @param frameX Frame X candiate for SCC calculation
+   * @param frameY Frame Y candidate for SCC calculation
+   * @param mode   An instance of {@link AstCorrelation.Mode}
+   * @return False if AstCorrelation.Mode is set to AllObs and any of the vectors in compared frames contains NaNs, otherwise True.
+   * @throws IllegalArgumentException When the {@link AstCorrelation.Mode} is set to AllObs and any of the vectors contains NaN
+   */
+  private static void checkCorrelationDoable(final Frame frameX, final Frame frameY, final AstCorrelation.Mode mode)
+          throws IllegalArgumentException {
+    if (!AstCorrelation.Mode.AllObs.equals(mode)) return;
+
+    final Vec[] vecsX = frameX.vecs();
+    final Vec[] vecsY = frameY.vecs();
+
+    assert vecsX.length == vecsY.length;
```

Comment:
should this really be an assert? is this checked somewhere else so it will never fail here? if user input can make this assert fail it should be an if () with an exception explaining the error to the user

Reply:
Truth be told, this rule should not even be applied. Having a non-square matrix is perfectly fine here. I'll add a test.

Reply:
Resolved in https://github.com/h2oai/h2o-3/pull/4173/commits/43759b10b4a76748073720233d94c1591fe47837 - non-squared correlation matrix is a perfectly fine approach. Thank you.

Also:
- Moved the check `isCorrelationDoable` after the null checks not to allocate any resources before this check in vain.
- The `isCorrelationDoable` newly checks the frames given for correlation calculation are not empty (don't have 0 columns).